syntax = "proto3";

package ChatProtocol;

message Packet {
    enum PacketType {
        // oneof enum값과 일치 시키기위한 temp enum값(proto3는 enum이 0부터시작)
        TEMP1 = 0;
        TEMP2 = 1;
        LOGIN_REQUEST = 2;
        LOGIN_REPLY = 3;
        ROOM_CREATE_REQUEST = 4;
        ROOM_CREATE_REPLY = 5;
        SEND_CHAT_REQUEST = 6;
        SEND_CHAT_REPLY = 7;
        NOTI_RECV_CHAT = 8;
        ROOM_JOIN_REQUEST = 9;
        ROOM_JOIN_REPLY = 10;
        SHOW_ROOM_LIST_REQUEST = 11;
        SHOW_ROOM_LIST_REPLY = 12;
        PACKET_MAX = 13;
    }
    
    PacketType type = 1;
    oneof packet_msg {
        LoginRequest login_request = 2;
        LoginReply login_reply = 3;
        RoomCreateRequest room_create_request = 4;
        RoomCreateReply room_create_reply = 5;
        SendChatRequest send_chat_request = 6;
        SendChatReply send_chat_reply = 7;
        NotiRecvChat noti_recv_chat = 8;
        RoomJoinRequest room_Join_request = 9;
        RoomJoinReply room_Join_reply = 10;
        ShowRoomListRequest show_room_list_request = 11;
        ShowRoomListReply show_room_list_reply = 12;
    }
}

enum ErrorType {
    ERROR_NONE = 0;
    ERROR_EXIST = 1; // TODO : 에러 타입 다양하게
}


//LOGIN_REQUEST
message LoginRequest {
    string nickname = 1;
}
//LOGIN_REPLY
message LoginReply {
    ErrorType err = 1;
    int32 ID = 2;
}

message RoomCreateRequest {
    string roomname = 1;
}

message RoomCreateReply {
    ErrorType err = 1;
    int32 room_number = 2;
}

message SendChatRequest {
    string chat = 1 ;
}

message SendChatReply {
    ErrorType err = 1;
}

message NotiRecvChat {
    string nickname = 1;
    string chat = 2;
}

message RoomJoinRequest {
    int32 room_number = 1;
}

message RoomJoinReply {
    ErrorType err = 1;
    int32 room_number = 2;
}

message ShowRoomListRequest {

}

message ShowRoomListReply {
    ErrorType err = 1;
    message RoomInfo {
        int32 number = 1;
        string name = 2;
    }
    repeated RoomInfo roomlist = 2;
}


